src/aiovodafone/api.py

```diff
@@ -118,10 +120,191 @@ async def _get_page_result(self, page: str) -> dict[Any, Any]:
     """Queries Vodafone Station running Technicolor firmware."""
 
+    def __init__(self, host: str, username: str, password: str) -> None:
+        super().__init__(host, username, password)
+        self.headers = {
+            "User-Agent": "Mozilla/5.0",
+            "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
+            "X-CSRF-TOKEN": "",
+            "X-Requested-With": "XMLHttpRequest",
+            "Referer": self.host,
+        }
+
+    async def _encrypt_string(
+        self, credential: str, salt: str, salt_web_ui: str
+    ) -> str:
+        """Calculates login hash from the password, the salt and the salt from the web UI.
+
+        Args:
+            credential (str): login password for the user
+            salt (str): salt given by the login response
+            salt_web_ui (str): salt given by the web UI
+
+        Returns:
+            str: the hash for the session API
+        """
+        _LOGGER.debug("Calculate hash")
+        a = hashlib.pbkdf2_hmac(
+            "sha256",
+            bytes(credential, "utf-8"),
+            bytes(salt, "utf-8"),
+            1000,
+        ).hex()[:32]
+        b = hashlib.pbkdf2_hmac(
+            "sha256",
+            bytes(a, "utf-8"),
+            bytes(salt_web_ui, "utf-8"),
+            1000,
+        ).hex()[:32]
+        return b
+
+    async def login(self) -> bool:
+        """Router login."""
+        _LOGGER.debug("Logging into %s", self.host)
+        _LOGGER.debug("Get salt for login")
+        page = self.host + "/api/v1/session/login"
+        reply = await self.session.post(
+            page,
+            headers=self.headers,
+            data={"username": self.username, "password": "seeksalthash"},
+        )
+        salt_response = await reply.json()
+        _LOGGER.debug("POST reply (%s)", page)
+
+        salt = salt_response["salt"]
+        salt_web_ui = salt_response["saltwebui"]
+
+        # Calculate hash
+        password_hash = await self._encrypt_string(self.password, salt, salt_web_ui)
+
+        # Perform login
+        _LOGGER.debug("Perform login")
+        page = self.host + "/api/v1/session/login"
+        reply = await self.session.post(
+            page,
+            headers=self.headers,
+            data={"username": self.username, "password": password_hash},
+        )
+        login_response = await reply.json()
+        _LOGGER.debug("POST reply (%s)", page)
+
+        if "error" in login_response and login_response["error"] == "error":
+            if login_response["message"] == "MSG_LOGIN_150":
+                raise AlreadyLogged
+            _LOGGER.error(await reply.text())
+            raise CannotAuthenticate
+
+        # Request menu otherwise the next call fails
+        _LOGGER.debug("Get menu")
+        now = int(time.time() * 1000)
+        page = self.host + "/api/v1/session/menu?_=" + str(now)
+        reply = await self.session.get(
```

Comment:
We should use a common `get` method

Reply:
I thought about it, too. The common method adds a CSRF token which is not needed in this implementation. Not sure how we want to handle the different parameters. Basically we would end up in different implementations for both. I could see how I can extract a generalized method.

Reply:
We can always have a bool to specify if CSRF should be appended or not. But at least we don't repeat the same code all over the libray.